### PR TITLE
Optimize trivial branches obscured by goto blocks on the fall-through path

### PR DESCRIPTION
**Pull request by ymanton**
This PR contains the following 2 commits:

1 - Teach Simplifier to detect conditional branches that have fallthrough
blocks that are goto blocks to the same target.

For example:
```
n223n     BBStart <block_18> (freq 1) (cold)
n210n     ifacmpeq --> block_17 BBStart at n219n ()
n208n       aload  <temp slot 5>[#529  Auto] [flags 0x7 0x0 ]
n209n       aconst NULL (X==0 X>=0 X<=0 )
n224n     BBEnd </block_18> (cold) =====

n227n     BBStart <block_19> (freq 1) (cold)
n226n     goto --> block_17 BBStart at n219n ()
n228n     BBEnd </block_19> (cold) =====
```
This complements the symmetrical case where the conditional branch
goes to the fallthrough block, which is already detected and removed by
Simplifier.

2 - Make conditional branches with a fallthrough goto block to the branch
target unconditional so that they can be removed by later passes.

### CODE REVIEW

File: compiler/optimizer/OMRSimplifierHelpers.cpp [hunk 1]
```diff
@@ -440,6 +440,26 @@ bool branchToFollowingBlock(TR::Node * node, TR::Block * block, TR::Compilation
    return true;
    }
 
+bool fallthroughGoesToBranchBlock(TR::Node *node, TR::Block *block, TR::Compilation *comp)
+   {
+   TR::Block *fallthroughBlock = block->getNextBlock();
+   if (fallthroughBlock == NULL ||
+       fallthroughBlock->isGotoBlock(comp) == false ||
```
Comment (hzongaro) [inline compiler/optimizer/OMRSimplifierHelpers.cpp:447]: May I ask you to change this line to `!fallthroughBlock->isGotoBlock(comp)`?

**Review by hzongaro — COMMENTED** [review]
I think these changes look good.  One cosmetic comment.

Also, I'm not sure why this change needs to be delivered as two commits.  May I suggest that you squash them?

File: compiler/optimizer/OMRSimplifierHelpers.cpp (hunk 1 above)
Comment (ymanton) [reply compiler/optimizer/OMRSimplifierHelpers.cpp]: Fixed, thanks.

**Comment by ymanton** [comment]
> Also, I'm not sure why this change needs to be delivered as two commits. May I suggest that you squash them?

Squashed, thanks.

**Comment by ymanton** [comment]
Added a potential fix for a BlockSplitter bug exposed by testing https://github.com/eclipse-openj9/openj9/pull/20127.

**Comment by vijaysun-omr** [comment]
If the WIP is taken off, I can review and run tests on this.

**Comment by vijaysun-omr** [comment]
jenkins build all

**Comment by vijaysun-omr** [comment]
The change looks fine to me, in that it handles a case that was probably exceedingly unlikely before an unrelated recent change was done. It does so under a conditional anyway, and so cases that used to work would presumably still work and this is only an improvement (from what was there before) for the cases that would have an issue.

**Pull request merged**